### ISSUE TITLE
Rethink of format options/configuration: Prepending options to the format URI

### ISSUE THREAD

**Issue by rsanger** [LibtraceTeam/libtrace#145]
TLDR: What are everyone's thoughts on being able to pass configuration options to a format by prepending those options before the format URI?
For example:
`./my_application  perpkt_threads=2,burst_size=10,ring:eth0`

Long version:
Most live formats have performance tuning options that would be nice to expose to libtrace users, now is a good time to reconsider how to pass configuration options through to specific libtrace formats.

Currently, Libtrace has 2 main functions and 2 main helper systems for configuring trace options:
**1.** The original `trace_config()` function, which uses the `TRACE_OPTION_<name>` flags
2. The friendly wrappers, `trace_set_<option_name>()` which have the argument type defined
**3.** New with the parallel API: `trace_set_configuration()` and `trace_set_configuration_file()` which take a string of key=value pairs and have a corresponding `trace_set_<name>()` function.
4. Environment string wrapper around `trace_set_configuration()` that looks for `FORMAT_CONF`, `LIBTRACE_CONF_<FORMAT>`, `LIBTRACE_CONF_<FORMAT_URI>` environment variables and passes them into `trace_set_configuration()`.

1 and 2 have the issue that every libtrace application needs to be updated to support new format options. 3 and 4 resolve this, however, are still a bit cumbersome to use.

@jacobvw has just added a new `coremap` option and, at my suggestion, added this as an option that could prepend the trace URI. For example, `coremap=[1:2:3]:int:eth0`.

This feels like a much more convenient way to pass these types of options into libtrace, so this issue is to figure out if and how we go about doing that in a standard manner.

I think this makes sense to build on top of the `trace_set_configuration()` logic. Then we could have comma-seperated list of key and value pairs followed by the format URI. For example:
`<key>=<value>,[<key>=<value>],[...],<format URI>`
`./my_application  perpkt_threads=2,burst_size=10,ring:eth0`

The first unknown key would be interpreted as the format URI. A typo of an option would result in a pretty obvious error that libtrace was trying to read an option as the format URI.

Or we could use a special character, such as a semi-colon (colon is already used enough), to end the option list:
`<key>=<value>,[<key>=<value>],[...];<format URI>`
`./my_application  perpkt_threads=2,burst_size=10;ring:eth0`

Right now options mainly take numerical values, but it is also worth considering if we need to keep any special characters available for future options.

Any thoughts on this and where we could run into issues?

I cannot think of any backward compatibility or other problems this would cause. My vote is in favour of doing this, or something like it.

**Comment by rsanger** [comment]
A solution is now in develop 6e55d85

The format is:
`<key>=<value>,...,<key>=<value>:<format URI>`
Where `format URI` is normally `<format>:<path>`

I was going to use semi-colon as the separator between options and URI, but had forgotten this already has meaning in most shells.

If a value requires commas, equals signs, or colons they can be enclosed in square brackets.
  - The code does a simple search for the trailing bracket
  - e.g. `coremap=[0,2], ... : <format URI>`

This logic also applies to `trace_set_configuration()` and the LIBTRACE_CONF environment variables.

Because all options are `<key>=<value>` and URIs are typically `<format>:<path>` it is easy to separate the two in the parsing logic.
If an option is invalid, the code falls back to parsing from this point onwards as a standard URI, this will account for complex file paths which contain the '=' signs.

**Comment by rsanger** [comment]
I've documented this on the [wiki](https://github.com/LibtraceTeam/libtrace/wiki/Passing-Options-to-Trace-Formats), it is detailed in the doxygen API documentation.
